programs/stake/src/stake_state.rs

```diff
@@ -553,9 +568,22 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
             }
             StakeState::Stake(meta, mut stake) => {
                 meta.authorized.check(signers, StakeAuthorize::Staker)?;
+                let stake_amount = self.lamports()?.saturating_sub(meta.rent_exempt_reserve); // can't stake the rent ;)
+                if stake_amount < MINIMUM_STAKE_DELEGATION {
+                    // Since (1) delegation uses the full balance (minus rent_exempt_reserve), and
+                    // (2) stake accounts must be created with a minimum balance of
+                    // rent_exempt_reserve + MINIMUM_STAKE_DELEGATION, then we can be sure that the
+                    // stake_amount will always be at least MINIMUM_STAKE_DELEGATION.
```

Comment:
it seems that there's counter example: https://explorer.solana.com/address/21mQHA6YZg5iX9KAsro4sk76D8jWwov1K64uKLDChV6J

```
$ ./target/release/solana stake-account 21mQHA6YZg5iX9KAsro4sk76D8jWwov1K64uKLDChV6J
Balance: 0.00228288 SOL
Rent Exempt Reserve: 0.00228288 SOL
Delegated Stake: 0 SOL
Active Stake: 0 SOL
Activating Stake: 0 SOL
Stake activates starting from epoch: 181
Delegated Vote Account Address: 9c5bpzVRbfsYY2fannb4hyX5CJUPg3BfH2cL6sR7kJM4
Stake Authority: 488StDvQbiDZeLSAqmpBz2zvjGe1LaAqwZckS21vufBP
Withdraw Authority: 488StDvQbiDZeLSAqmpBz2zvjGe1LaAqwZckS21vufBP
```

Reply:
as a quick context, there's long-tall tragedy of the ugly current state of code and live data for stake accounts. ;)

Reply:
i haven't dug in detail. but fully-deactivated withdraw allow this condition to occur?

Reply:
Oh... Uh... I'm not sure. @CriesofCarrots or @joncinque do you know?

Reply:
Even from a deactivated stake, you can withdraw up to reserve + 1 lamport. There's currently code to ensure that the reserve is kept, which was correctly updated: https://github.com/solana-labs/solana/blob/86d465c53137978b464e01c1e1155b9f58b166cb/programs/stake/src/stake_state.rs#L799

It was possible to split down to just reserve before, but that was fixed.  This account is probably a holdover from back then, but thankfully it's possible to withdraw all the lamports out of it

Reply:
Great! Marking this as resolved. Thanks!

Reply:
What Jon said :)

I believe the only operations this account will allow would be to withdraw to zero, or add lamports >= 1. @brooksprumo , if this one particular stake account tried to withdraw something less than 0.00228288 SOL, which error would it return?

@ryoqun , any idea how many stake accounts are in this state? Are there any that still exist with rent-exempt reserve of greater than 0.00228288 SOL?

Reply:
@CriesofCarrots As long as the account is not `Uninitialized`, then the error will be `InsufficientFunds`
https://github.com/solana-labs/solana/blob/a79819978cb507a8f8f49bbf6e2b37a13704ccea/programs/stake/src/stake_state.rs#L806

If the account _is_ `Uninitialized`, then any amount can with withdrawn (up to the actual balance, of course).

Can these old accounts ever be `Uninitialized`?

Reply:
Cool, thanks for tracing that for me. Just wanted to check that we couldn't end up with the new error somehow.
Yeah, the weird accounts could be `Uninitialized`. But then their only viable actions would be to keep withdrawing until they reach zero, or deposit until they reach rent-exemt-minimum+1, so no change in behavior.

Reply:
> Even from a deactivated stake, you can withdraw up to reserve + 1 lamport. There's currently code to ensure that the reserve is kept,

well, i haven't tested locally, but it's possible to create 0-lamports delegated stake account if i read correctly?

https://github.com/solana-labs/solana/blob/a79819978cb507a8f8f49bbf6e2b37a13704ccea/programs/stake/src/stake_state.rs#L754-L768

fully-deactivated stake account should fall into this match arm and return the tuple `(<lockup>, 0+rent_exempt_reserve, false)`

then

https://github.com/solana-labs/solana/blob/a79819978cb507a8f8f49bbf6e2b37a13704ccea/programs/stake/src/stake_state.rs#L802-L806

this `if` shouldn't be triggered  because `withdraw_amount + rent_exempt_reserve > account_balance` => `false` (note: not `>=` here) for all withdrawn lamports but the rent_exempt_reserve


https://github.com/solana-labs/solana/blob/d5dec989b93fc94c9416a60193e52e8b68bd7133/programs/stake/src/stake_state.rs#L554-L568

then, this `stake_lamports` just be `0` without causing errors when redelegating such a stake account later, as this is _saturating_ subtraction.

thus, this can introduce unintended behavior change maybe by adding new `if` in this pr?

Reply:
@brooksprumo yeah, initialize works as intended. however, stake accounts can be recycled after fully deactivated bypassing the `initialize` altogether like this:

here's more detailed steps of https://github.com/solana-labs/solana/pull/22663#discussion_r802632124

1. create a normal stake account with good funding
2. delegate it to some validator
3. deactivate it
4. withdraw all but `rent_exempt_reserve`
5. delegate it to some validator again

i've just confirmed locally that you can still create 0-stake active stake accounts this way.

Reply:
@CriesofCarrots 

> @ryoqun , any idea how many stake accounts are in this state?

really small (10 accounts):

```
$ less mainnet-beta-272-2.ledger-tool-inflation-validation-minimum-stake-delegation.stderr1 | grep -E 'account|deacti' | grep -E 'N/A|account' | grep -B1 N/A | grep -v -- --
    account: "12FFUP9zeUTDYvYkv8md2mJHsFD27pUfHFZCW9XTRHba",
    deactivation_epoch: "N/A",
    account: "21mQHA6YZg5iX9KAsro4sk76D8jWwov1K64uKLDChV6J",
    deactivation_epoch: "N/A",
    account: "28AW2otHAuyYXfMrLqLqiKKrA4csMXnFBXEzuzepNC7w",
    deactivation_epoch: "N/A",
    account: "2CSZmTsaTon1TMDSnuJ7vZ3ixM27DgCUDPaEutwosXtg",
    deactivation_epoch: "N/A",
    account: "6V3depRzCh49de2GaxMJRLa9oeZjhe83mrphRaLP42jp",
    deactivation_epoch: "N/A",
    account: "8ZwXhv54DfTWB96PvK3tCyiXRXuGCUbEGBXz9YNrYNud",
    deactivation_epoch: "N/A",
    account: "BqaFpYxiy6MH4Z4wPVh32Tvo6M7kR4dddw9WMyGiu7Fe",
    deactivation_epoch: "N/A",
    account: "Cs9pJKWUPuhBibvbSFizYVGqJDe6enkD2ountgBNi9SN",
    deactivation_epoch: "N/A",
    account: "HrJA7JMzZnf2YmqWQz5N2ke2MpgzCAFtyzrA2yGsAh64",
    deactivation_epoch: "N/A",
    account: "HtTouwfJPWudZj2gRY7Q6RoyfBi25EV5sYV6Q65XZNo4",
    deactivation_epoch: "N/A",
```

however, there's several accounts (~178), which could be 0-lamports delegated stakes at any time (currently residing at the step 4 from https://github.com/solana-labs/solana/pull/22663#discussion_r804319877)

```
$ less mainnet-beta-272-2.ledger-tool-inflation-validation-minimum-stake-delegation.stderr1 | grep -E 'deacti' | grep -v N/A | wc -l
178
```

>  Are there any that still exist with rent-exempt reserve of greater than 0.00228288 SOL?

yeah, there is some (~0.1%).

```
# stake account count by its rent-exempt reserve field for mainnet-beta epoch 272
$ cat mainnet-beta-272-2.csv | grep Stake11111111111111111111111111111111111111 | tr , \  | awk '{print $12}' | sort | uniq -c
 691171 2282880
    673 28786560

```

here's a pointer (https://github.com/solana-labs/solana/pull/13461) for those who don't know the history. :)

now that realloc and sysvars are available, i think it's time to clean up this mess finally? ;) I'm bit reluctant though, considering its benefits.

Reply:
@ryoqun Thanks for the detailed issue and steps to reproduce. For this PR, I'm not sure what the correct behavior should be. Here's the fn that validates the stake amount:

https://github.com/solana-labs/solana/blob/a79819978cb507a8f8f49bbf6e2b37a13704ccea/programs/stake/src/stake_state.rs#L826-L846

And here's `delegate()` from master:

https://github.com/solana-labs/solana/blob/c078ca3fb3e203dbd4f3a9556166695e8223cea7/programs/stake/src/stake_state.rs#L530-L569

Should I basically (1) re-write this comment, and (2) replace the `Err` with a `warn!()`?

https://github.com/solana-labs/solana/blob/a79819978cb507a8f8f49bbf6e2b37a13704ccea/programs/stake/src/stake_state.rs#L834-L843

Then I think that would retain the current behavior.

Then I could go the full feature-gate for a subsequent PR that puts the error back. Would that work?

Reply:
If in doubt, just gate it.  No sense bending over backwards avoiding it just to fork the cluster on a corner case anyway

Reply:
> Then I could go the full feature-gate for a subsequent PR that puts the error back.

yeah. sounds good. the additional bonus would be add tests in this pr to validate the current behavior for the newly-found minimum stake delegation loophole (i.e. withdraw & re-delegate with fully-deactivated & only-just-enough-rent-exempt-reserve) so that the subsequent feature gate pr's behavior change should be validated just by tweaking the `assert!(..)`s in the tests.

> Should I basically (1) re-write this comment

Yeah.

> and (2) replace the `Err` with a `warn!()`?

Well, how much `warn!()` would be beneficial to node operators? i think this just scares them for no good reason because those `warn!()` are just not actionable for them. also, this will allow little annoyance of allowing log spamming by attackers.

Reply:
Done!